src/rpc/utils.rs

```diff
@@ -0,0 +1,59 @@
+use openssl::hash;
+use serde::ser::{self, Serialize};
+use serde::de::{self, Deserialize};
+
+use error;
+
+
+pub(crate) fn sha1_bytes(parts: &[&[u8]]) -> error::Result<Vec<u8>> {
+    let mut hasher = hash::Hasher::new(hash::MessageDigest::sha1())?;
+    for part in parts {
+        hasher.update(part)?;
+    }
+
+    let bytes = hasher.finish2().map(|b| b.to_vec())?;
+
+    Ok(bytes)
+}
+
+
+#[derive(Debug)]
+pub enum EitherRef<'a, T: 'a> {
+    Ref(&'a T),
+    Owned(T),
+}
```

Comment:
isn't this `Cow`?

Reply:
In short: we need a certain degree of dependency between the reference and the owned variants, but not too much.`Cow` is too restrictive while `A` and `B` in `Either<A, B>` are completely independent from each other, so that even `type EitherRef<'a, T> = Either<&'a T, T>;` won't help much.

Here is a very long explanation which was unconsciously typed down before the short summary above and now I don't feel like I can delete it. Sorry for being too wordy! :D
<details>
  <summary>Long text</summary>

* There is a blanket
  ```rust
  impl<T> ToOwned for T
      where T: Clone
  {
      type Owned = T;
      fn to_owned(&self) -> T {
          self.clone()
      }

      // Irrelevant for us
      // ...
  }
  ```
  implementation - if we keep using just `Cow<'a, T>`, the trait bounds will force us to require `T: Clone` everywhere just to fulfill the type correctness when the semantics of `Message` [de]serialization doesn't even need cloning anywere;
* So we need to get rid of the `Clone` bound but preserve the borrowing ability. The only safe way to do that is disallowing creating owned values from borrowed ones (correct me if I'm wrong here). Let's show that we can't do this.
  Assume that we can do it. This boils down to solving the problem:
  Find such types `U != !` and `V != !` (empty types are useless for us), so that:
  - for any value `u` of type `U`: there exists a value `v` of type `V`: so that `&u` can be borrowed as `&v`;
  - for any type `V: ToOwned`: there doesn't exist any value in `V::Owned`.

  Both points above, combined, are equivalent to
  - all of `U: Borrow<V>`, `V: ToOwned` and `V::Owned = !` must be true.

  That implies `!: Borrow<V>` as the only solution. So `U = !` as the conclusion. Contradiction with proposition.
* Okay, so how about non-`Cow` solutions? `Either` comes to me as the only one (please propose if there are more than I know of). The downside of it was the [de]serialization: it behaves like any enum with two variants with a single field in each would behave, when we need a special treatment of owned vs. borrowed values.

Unable to find a ready-to-use solution for this case (which is, I guess, rather obscure if community hasn't come up with it yet), I rolled out mine which you can see now. If you know something better, please tell me.

Thank you for reading up to this point! :D

</details>